src/easylink/implementation.py

```diff
@@ -111,15 +116,82 @@ def _validate_expected_steps(self, logs: list[str]) -> list[str]:
             )
         return logs
 
-    def _validate_container_exists(
+    def _download_and_validate_image(
         self, logs: list[str], images_dir: str | Path
     ) -> list[str]:
-        """Validates that the container to run exists."""
+        """Downloads the image if required and validates it exists.
+
+        If the image does not exist in the specified images directory, it will
+        attempt to download it.
+        """
         # HACK: We manually create the image path here as well as later when writing
         # each implementations Snakefile rule.
         image_path = Path(images_dir) / self.singularity_image_name
+        expected_md5_checksum = self._metadata.get("md5_checksum", None)
+        record_id = self._metadata.get("zenodo_record_id", None)
+        if image_path.exists():
+            self._handle_conflicting_checksums(
+                logs, image_path, expected_md5_checksum, record_id
+            )
+        else:
+            if not record_id:
+                logs.append(
+                    f"Image '{str(image_path)}' does not exist and no Zenodo record ID "
+                    "is provided to download it."
+                )
+            if not expected_md5_checksum:
+                logs.append(
+                    f"Image '{str(image_path)}' does not exist and no MD5 checksum "
+                    "is provided to verify from the host."
+                )
+            if not record_id or not expected_md5_checksum:
+                return logs
+            download_image(
+                images_dir=images_dir,
+                record_id=record_id,
+                filename=self.singularity_image_name,
+                md5_checksum=expected_md5_checksum,
+            )
         if not image_path.exists():
-            logs.append(f"Container '{str(image_path)}' does not exist.")
+            logs.append(
+                f"Image '{str(image_path)}' does not exist and could not be downloaded."
+            )
+        return logs
+
+    @staticmethod
+    def _handle_conflicting_checksums(
+        logs: list[str],
+        image_path: Path,
+        expected_md5_checksum: str | None,
+        record_id: str | None,
+    ) -> list[str]:
+        # TODO: Strengthen the following logic to better handle image updates.
```

Comment:
We all thought for a while about how to best handle this and this seems like a decent compromise. I'm open to suggestions if folks find this is troublesome in practice.